### ISSUE TITLE
Pass all the environment variables to the functions

### ISSUE THREAD

**Issue by FilipPyrek** [bboure/serverless-appsync-simulator#22]
**Is your feature request related to a problem? Please describe.**
When Lambda function is deployed, the environment variables like `AWS_REGION` are implicitly set by AWS itself. So for example in NodeJS the `aws-sdk` can extract the region value from the environment variable. But when locally using `serverless-appsync-simulator`, only environment variables defined in `serverless.yml` file are passed to the function and thus the SDK throws an error, because it can't find the region in the environment variable.

Code example:

How I would like the code to look like 🙂 :
```typescript
import AWS from 'aws-sdk'

const dynamodb = new AWS.DynamoDB()
const client = await dynamodb.getItem({
  TableName: process.env.DYNAMODB_TABLE_NAME,
  Key: {
    id: {
      S: `${client.id}`
    }
  }
}).promise()
```

How it has to be done now 😢 :
```typescript
import AWS from 'aws-sdk'

const dynamodb = new AWS.DynamoDB({ region: 'ap-southeast-1' }) // region has to explicitly defined here
const client = await dynamodb.getItem({
  TableName: process.env.DYNAMODB_TABLE_NAME,
  Key: {
    id: {
      S: `${client.id}`
    }
  }
}).promise()
```

**Describe the solution you'd like**
Pass all the environment variables to the function. 

During local development we are using [`env-cmd`](https://www.npmjs.com/package/env-cmd) file through which we mock all the environment variables. So for example we define there `AWS_REGION` value.

**Describe alternatives you've considered**
I tried to set `AWS_REGION` environment variable in `provider.environment`, but later during deployment AWS CloudFormation throws and error, because it's a reserved environment variable name.

**Comment by bboure** [comment]
@FilipPyrek  I solved this kind of problem using https://github.com/colynb/serverless-dotenv-plugin

I just have a `.env.local` file where I set:
````
AWS_REGION=local
````

In that file you can override any env var from `provider.environment` or define additional ones (like `AWS_REGION`), or just stage specific values.

Then I run the simulator with,  `sls offline start -s local --env local`

`-s` is the serverless stage
`--env` tells `dotEnv` which `.env` file to use (in my case that is `.env.local`).

This way, when deploying to another stage the `AWS_REGION` environment is never set and you don't get the a CloudFormation error, because the .env file is not used, or another one without that value is used.

Passing all the local environment variables to the functions is not done on purpose to avoid passing more than necessary and avoid clashes.

**Comment by FilipPyrek** [comment]
Even tough I understand you point @bboure I don't think I's a way in our case at @purple-technology because we already have some workflow in place with `.env` files and when I tried dynamically generate configuration for the serverless-dotenv-plugin I didn't succeed because the plugin is not processing the the incoming configuration values with `serverless.variables.getValueFromSource` thus doing stuff like `dotenv:  ${file(macro.js):dotenvConfig}` is not possible...

What about doing a compromise and pass into the functions only environment variables with prefix `AWS_`? 

https://github.com/bboure/serverless-appsync-simulator/blob/master/src/getAppSyncConfig.js#L78-L81

Sample code of the variables filter:
```javascript
Object.keys(process.env).reduce(
  (acc, varName) =>
    /^AWS_[A-Z0-9_]+$/.test(varName)
      ? { ...acc, [varName]: process.env[varName] }
      : acc,
  {}
)
```

**Comment by bboure** [comment]
@FilipPyrek I am not sure I understood your sue case. 
Are you having dynamically generated env vars through `macro.js` ? 
(If that is the case, I would be curious to see how you guys do that. That could be useful :)) 

In any case, I can offer you another compromise. Or maybe a couple of them:

1) Add a `custom.appsync-simulator.lambda.loadLocalEnv` option.
If `true`, the plugin will load all local variables into the lambdas.
Default would be `false`


2) Add a `custom.appsync-simulator.lambda.environment` option.
Contains an array of **additional** environment variables to be injected only when the simulator is running.

**Comment by FilipPyrek** [comment]
Our use case:
- we use monorepo thus we have multiple SLS services in one repo
- we use single `.env` file located in the root directory of the repo
- we load the `.env` with [env-cmd](https://www.npmjs.com/package/env-cmd) plugin. Example: `$ env-cmd -f ../.env serverless deploy`

My idea of using [serverless-dotenv-plugin](https://github.com/colynb/serverless-dotenv-plugin) as you suggested in our monorepo:
- we would use [serverless-dotenv-plugin](https://github.com/colynb/serverless-dotenv-plugin) only for `serverless offline`
- but there is no if-else in serverless.yml
- thus we would have to create a "macro" as I call it, which would be a simple JS script which would be loaded as follows: `dotenv: ${file(macros.js):dotenvConfig}`
- this way,  the `dotenvConfig` function in `macros.js` file would return different configuration in different situations
- in case of `serverless offline`, it would return configuration with `basePath: ../` - where the .env file is located
- in all other cases e.g deployment, it would return empty configuration - thus the plugin wouldn't be able to find the `.env` file - thus the plugin wouldn't touch the environment variables at all.

**Comment by FilipPyrek** [comment]
Perfect, I love you proposals - actually at first I was thinking about point 1. but I thought that maybe it would be too much interference into the plugin so I decided to propose the `AWS_` prefix.

So I vote for the option 1. I think it's the best solution for this problem.

**Comment by bboure** [comment]
> So I vote for the option 1. I think it's the best solution for this problem.

You are more than welcome to open a PR if you feel like it :)

**Comment by FilipPyrek** [comment]
As I'm working on the PR for serverless-appsync-simulator now, would you mind switching from `yarn` to `npm` @bboure ?

**Comment by FilipPyrek** [comment]
@bboure I'm getting error `Cannot find module 'extend'` from this line: https://github.com/aws-amplify/amplify-cli/blob/master/packages/amplify-appsync-simulator/src/server/subscription/mqtt-server/server.ts#L100

When got rid of the `extend` function in the file and replaced it with spread operator  ` { ...DEFAULT_OPTIONS, ...options }` as a hotfix it started to work.

But I'm wondering how is it possible that published version of `serverless-appsync-simulator` works even though `amplify-appsync-simulator` seems to be somehow broken? 😃 
Because it's not in `package.json`: https://github.com/aws-amplify/amplify-cli/blob/master/packages/amplify-appsync-simulator/package.json

**Comment by bboure** [comment]
Hum @FilipPyrek , Good catch!
I have seen that error before too while working on the package, and could not figure out what the issue was.
I thought it was an issue with babel (I guess I was mislead by the word `extend`).

I think that amplify released a new version of amplify-simulator recently.
Could you try to
`yarn upgrade --latest`
There is [a fix](https://github.com/aws-amplify/amplify-cli/pull/4203) I would like to include anyway.

If that does not work, maybe a dirty workaround would be to include that module in our package.
Or open an issue/PR in `amplify-cli`.

**Comment by bboure** [comment]
> would you mind switching from yarn to npm @bboure 

Haa, the eternal debate [npm vs yarn](https://www.google.com/search?newwindow=1&sxsrf=ALeKk01YNnfdNd_ZH2qGh-TRvSlxuTMVaQ%3A1589565627241&ei=u9i-XqyVDo64kwXr_7j4DA&q=npm+vs+yarn&oq=npm+vs+yarn&gs_lcp=CgZwc3ktYWIQAzIFCAAQkQIyAggAMgIIADICCAAyAggAMgIIADICCAAyAggAMgIIADICCAA6BAgAEEc6BAgjECc6BAgAEEM6BwgAEBQQhwJQ-dYCWKfdAmDd3gJoAHADeACAAYsCiAHhCJIBBTAuMi4zmAEAoAEBqgEHZ3dzLXdpeg&sclient=psy-ab&ved=0ahUKEwjs3ru0ubbpAhUO3KQKHes_Ds8Q4dUDCAw&uact=5)

I have been wondering recently if I should switch back.

I guess that in this project, `npm` should be ok. 
But maybe I'll keep that for another PR if you don't mind.

**Comment by FilipPyrek** [comment]
I seems that when I install dependencies with `yarn`, it works. But when I install dependencies with `npm` I get the `Cannot find module 'extend'` error. So for now I will stick with Yarn 😄 

But I will submit the PR in amplify-cli anyway.

**Comment by bboure** [comment]
@FilipPyrek reason is probably because there is no `package-lock.json` file, so it installs the latest version by default.
With yarn, it sticks to the frozen version.
Thanks!

**Comment by bboure** [comment]
Fixed by #24
